docs/adrs/adr007-events-validation.md

```diff
@@ -0,0 +1,142 @@
+# ADR007: Events validation
+
+Status: accepted
+
+<!-- These documents have names that are short noun phrases. For example, "ADR001: Deployment on Ruby on Rails 3.0.10" or "ADR009: LDAP for Multitenant Integration" -->
+
+## Context
+
+Widgets are using Matrix events that are defined in Matrix specification and as well custom events that are
+created and used by other custom matrix clients or widgets.
+
+Matrix specification states that events should be considered as untrusted data. This applies both to Matrix and
+custom events.
+
+[MSC 2801][msc2801] provides reasons why events may not match the specification and the conclusion that
+clients must treat all event data as untrusted.
+
+Matrix events are described by matrix specification with [event schemas][eventSchemas]. This should be used as
+a reference to implement validation logic of the events used.
+
+Widgets may also load data from other data sources: localstorage, RTC channel, HTTP APIs. In general,
+we could not consider these events and data to be of needed schemas all the time and therefore must validate.
+
+<!--
+This section describes the forces at play, including technological, political, social, and project local. These forces are probably in tension, and should be called out as such. The language in this section is value-neutral. It is simply describing facts. -->
+
+## Decision
+
+<!-- This section describes our response to these forces. It is stated in full sentences, with active voice. "We will ..." -->
+
+We validate all events and data that are coming into widgets from different sources:
+
+For Matrix events (both from Matrix specification and custom widget events):
+
+- `type` has expected event type, example: `m.room.name`, `net.nordeck.whiteboard`
+- `content` to be an `object` and to be defined
+- `content` is validated to match the schema
+
+For events and data from other data sources: we apply validation of data to match the schema.
+
+Schema based validation includes:
+
+- required fields are listed
+- optional and required fields values have the expected data types
+- fields primitive values match expected grammar: dates in ISO 8601, MXC URIs, whitelist values
+- fields object data is also validated according to schema
+
+We use several ways to validate incoming data.
+
+### Plain code
+
+Validation logic could be written in plain code when it is not complicated or we don't want to use some library as external dependency.
+
+Example that validates that event has `m.relates_to` in the content:
```

Comment:
```suggestion
Example for validating that an event has an [`m.relates_to` object](https://spec.matrix.org/v1.13/client-server-api/#forming-relationships-between-events) in the content:
```

We could additionally permalink to the real place in the code for this. Is this a good idea?
`https://github.com/nordeck/matrix-widget-toolkit/blob/20d3c085c65ed063577b4b318aefb4b1610e3d8b/packages/api/src/api/extras/relatesTo.ts#L91-L121`

Reply:
yes, good Idea to provide more context for that. Links are added to two examples in adr.